src/validation/schema.js

```diff
@@ -322,13 +378,203 @@ const validateEntityFields = (defs, def) =>
+        immutable.fromJS({
+          loc: def.name.loc,
+          entity: def.name.value,
+          message: `@imports directive only allowed on '${RESERVED_TYPE}' type`,
+        }),
+      )
+    : List()
+
+const importDirectiveTypeValidators = {
+  StringValue: (_def, _directive, _type) => List(),
+  ObjectValue: (def, directive, type) => {
+    let errors = List()
+    if (type.fields.length != 2) {
+      return errors.push(
+        immutable.fromJS({
+          loc: directive.name.loc,
+          entity: def.name.value,
+          message:
+            'Imported type must be one of "Name", { name: "Name" }, or { name: "Name", as: "Alias" }',
+        }),
+      )
+    }
+    return type.fields.reduce((errors, field) => {
+      if (!['name', 'as'].includes(field.name.value)) {
+        return errors.push(
+          immutable.fromJS({
+            loc: directive.name.loc,
+            entity: def.name.value,
+            message: `@imports type object field '${field.name.value}' invalid,  may only be one of: [name, as]`,
+          }),
+        )
+      }
+      if (field.value.kind != 'StringValue') {
+        return errors.push(
+          immutable.fromJS({
+            loc: directive.name.loc,
+            entity: def.name.value,
+            message: '@imports type object fields [name, as] must be strings',
+          }),
+        )
+      }
+      return errors
+    }, errors)
+  },
+}
+
+const validateImportDirectiveType = (def, directive, type) => {
+  return importDirectiveTypeValidators[type.kind]
+    ? importDirectiveTypeValidators[type.kind](def, directive, type)
+    : List().push(
+        immutable.fromJS({
+          loc: directive.name.loc,
+          entity: def.name.value,
+          message:
+            'Imported type must be one of "Name", { name: "Name" }, or { name: "Name", as: "Alias" }',
+        }),
+      )
+}
+
+const validateImportDirectiveArgumentTypesIsValid = (def, directive, argument) => {
+  if (argument.value.kind != 'ListValue') {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: '@imports directive argument: `types` must be an list',
+      }),
+    )
+  }
+
+  return argument.value.values.reduce(
+    (errors, type) => errors.concat(validateImportDirectiveType(def, directive, type)),
+    List(),
+  )
+}
+
+const validateImportDirectiveArgumentFromIsValid = (def, directive, argument) => {
+  if (argument.value.kind != 'ObjectValue') {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: '@imports directive argument: `from` must be an object',
+      }),
+    )
+  }
+
+  if (argument.value.fields.length != 1) {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: `@imports directive argument: 'from' must have one field of id or name`,
```

Comment:
`@imports directive argument:` -> `@import argument`

`field of id or name` -> improve to ```must have an `id` or `name` field```